### PR TITLE
Add unnecessary_stateful_widgets

### LINKED ISSUE #3683: lint: StatefulWidget could be a StatelessWidget
Occasionally, I see Flutter code where people use StatefulWidgets even though a StatelessWidget would have been sufficient. Example:

```dart
class MyWidget extends StatefulWidget {
  const MyWidget({super.key});

  @override
  State<MyWidget> createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  @override
  Widget build(BuildContext context) {
    return Container();
  }
}
```

The `State` class doesn't store any state in member variables and it also doesn't use any of the properties provided by `Sate` (e.g. `mounted`). This could easily be refactored to be a more concise `StatelessWidget`:

```dart
class MyWidget extends StatelessWidget {
  const MyWidget({super.key});

  @override
  Widget build(BuildContext context) {
    return Container();
  }
}
```

I wonder if it would be possible to have a lint for this?

Of course, from a functional perspective using a StatefulWidget here is totally fine, the additional performance overhead of a StatefulWidget over a StatelessWidget is probably neglectable. However, using a StatelessWidget where appropriate leads to more compact, concise, and readable code and is more idiomatic, IMO.

### LINKED ISSUE #58868: lint: StatefulWidget could be a StatelessWidget
Occasionally, I see Flutter code where people use StatefulWidgets even though a StatelessWidget would have been sufficient. Example:

```dart
class MyWidget extends StatefulWidget {
  const MyWidget({super.key});

  @override
  State<MyWidget> createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  @override
  Widget build(BuildContext context) {
    return Container();
  }
}
```

The `State` class doesn't store any state in member variables and it also doesn't use any of the properties provided by `Sate` (e.g. `mounted`). This could easily be refactored to be a more concise `StatelessWidget`:

```dart
class MyWidget extends StatelessWidget {
  const MyWidget({super.key});

  @override
  Widget build(BuildContext context) {
    return Container();
  }
}
```

I wonder if it would be possible to have a lint for this?

Of course, from a functional perspective using a StatefulWidget here is totally fine, the additional performance overhead of a StatefulWidget over a StatelessWidget is probably neglectable. However, using a StatelessWidget where appropriate leads to more compact, concise, and readable code and is more idiomatic, IMO.

### PR DESCRIPTION
**Pull request by a14n**
Fixes dart-lang/sdk#58868

### CODE REVIEW

File: test_data/rules/unnecessary_stateful_widgets.dart
Comment (goderbauer) [inline test_data/rules/unnecessary_stateful_widgets.dart:70]: What's the reasoning behind marking this as OK? I think, naively I would have marked this with the lint as well...

File: test_data/rules/unnecessary_stateful_widgets.dart
Comment (goderbauer) [inline test_data/rules/unnecessary_stateful_widgets.dart:60]: This should really be linted as well since the state is not really holding on to any state...

Comment (goderbauer) [reply test_data/rules/unnecessary_stateful_widgets.dart:60]: But, of course, if a member `i` holds on to anything stateful (say a FocusNode or a ScrollController) it shouldn't lint. I guess, we probably can't differentiate between these cases?

**Review by goderbauer — COMMENTED** [review]
How does this do when run over the flutter repository? Does it find anything? Any false alarms?

File: test_data/rules/unnecessary_stateful_widgets.dart
Comment (a14n) [reply test_data/rules/unnecessary_stateful_widgets.dart:70]: Removing the state would be a breaking change as someone could have referenced this type outside. For instance in flutter package `PopupMenuButtonState` is one of these cases (and the reason why I added this exception).

File: test_data/rules/unnecessary_stateful_widgets.dart
Comment (a14n) [reply test_data/rules/unnecessary_stateful_widgets.dart:60]: I started with only a rule to catch usage of `setState` but on the flutter codebase I face several false positives with State having fields. As you mentioned it's not easy to catch what field is stateful.

**Comment by a14n** [comment]
> How does this do when run over the flutter repository? Does it find anything?

On packages/flutter : https://github.com/flutter/flutter/pull/112296

> Any false alarms?

The exceptions I added in this linter PR is the result of a too broaden application.

File: test_data/rules/unnecessary_stateful_widgets.dart
Comment (a14n) [reply test_data/rules/unnecessary_stateful_widgets.dart:60]: If i remove the absence of fields in state condition, here's the list of diagnotics

```
   info • Unnecessary StatefulWidget • packages/flutter/lib/src/material/about.dart:530:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/lib/src/material/bottom_sheet.dart:67:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/lib/src/material/bottom_sheet.dart:350:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/lib/src/material/dropdown.dart:95:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/lib/src/widgets/fade_in_image.dart:61:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/test/material/reorderable_list_test.dart:1782:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/test/widgets/animated_list_test.dart:494:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/test/widgets/heroes_test.dart:162:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/test/widgets/inherited_model_test.dart:55:7 • unnecessary_stateful_widgets
   info • Unnecessary StatefulWidget • packages/flutter/test/widgets/slivers_block_global_key_test.dart:11:7 • unnecessary_stateful_widgets
```

A lot of them are not simple/impossible to make statelessWidget

**Comment by asashour** [comment]
How about checking if an override is a default one, so it is not preventing the lint from being triggered?

E.g. 

```dart
@override
void dispose() => super.dispose();
```

The relevant check is in [FlutterConvertToStatelessWidget](https://github.com/dart-lang/sdk/blob/cfddeee68e0cf7451a2eee1a9fbbff8a8d9c0b92/pkg/analysis_server/lib/src/services/correction/dart/flutter_convert_to_stateless_widget.dart#L228) assist.

**Comment by a14n** [comment]
> How about checking if an override is a default one, so it is not preventing the lint from being triggered?

Thanks for the suggestion.

This case should be catch by [unnecessary_overrides](https://dart-lang.github.io/linter/lints/unnecessary_overrides.html). Not sure it is worth adding complexity here to catch things like this.

This `unnecessary_stateful_widgets` could surely be really more complex to catch more cases. But is it worth it regarding the cost of maintenance and how many additionnal diagnotics could arise.

**Comment by goderbauer** [comment]
I added some comment over at: https://github.com/flutter/flutter/pull/112296#discussion_r981553217

I wonder if we should omit the lint if the State implementation is actually using functionality of the `State` class. E.g. if you access `State.context` or `State.mounted` or some such I think it is ok to be a stateful widget.

**Comment by a14n** [comment]
I updated the lint to have an exception if the WidgetState uses a member of its parent.

I think this PR is now ready from a functional point of view.

**Review by goderbauer — COMMENTED** [review]
LGTM for the cases in which the lint does and doesn't trigger (as documented in the `test_data`).

The lint implementation needs a review from somebody more knowledgable in the linter space... :)

**Comment by srawlins** [comment]
I want to double check that we're "accepting" the new lint request at https://github.com/dart-lang/linter/issues/3683. Then we'll review.

File: lib/src/rules/unnecessary_stateful_widgets.dart
Comment (srawlins) [inline lib/src/rules/unnecessary_stateful_widgets.dart:155]: Can you document what this method returns?

**Review by srawlins — APPROVED** [review]
Technically this looks good to me. The proposal has been accepted.

CC @pq @bwilkerson if you care to review from a technical perspective.

**Comment by devoncarew** [comment]
Thanks for the contribution! The code for the linter has since moved into the sdk repo (at https://github.com/dart-lang/sdk/tree/main/pkg/linter). If you think this PR is still relevant, consider re-creating it against the source there (see https://github.com/dart-lang/sdk/blob/main/CONTRIBUTING.md).

**Comment by Turskyi** [comment]
@devoncarew , @goderbauer , @srawlins I’m struggling to understand the current situation regarding this issue. If the unnecessary_stateful_widgets linter has moved to a new repository, why is the issue (https://github.com/dart-lang/linter/issues/3683) still open? 

If it hasn’t been addressed, why wasn’t the approved pull request merged during the year it remained pending, especially since it passed all tests and was approved?

It’s deeply concerning that contributions from the community—like this PR, which could have improved the framework—are effectively discarded due to delays in merging and the subsequent repository move. This discourages contributors who are genuinely invested in helping Flutter progress.

If the move to a new repository made the merge impossible, shouldn’t the team have proactively migrated this PR or implemented the changes themselves, given that it was already approved? Asking the contributor to recreate their work seems unfair and reflects poorly on the process.

I hope this feedback is taken constructively to avoid similar situations in the future and to better respect the time and efforts of community contributors.

**Comment by srawlins** [comment]
> I’m struggling to understand the current situation regarding this issue. If the unnecessary_stateful_widgets linter has moved to a new repository, why is the issue (https://github.com/dart-lang/linter/issues/3683) still open?

The issues have not yet been moved to the dart-lang/sdk repo. They will be.

> If it hasn’t been addressed, why wasn’t the approved pull request merged during the year it remained pending, especially since it passed all tests and was approved?

We are not always good about triaging this repository's pull request queue. Additionally, I see an un-addressed comment above.

> It’s deeply concerning that contributions from the community—like this PR, which could have improved the framework—are effectively discarded due to delays in merging and the subsequent repository move. This discourages contributors who are genuinely invested in helping Flutter progress.

Apologies for this. The Developer Experience team is actively working to improve our triage processes.

> If the move to a new repository made the merge impossible, shouldn’t the team have proactively migrated this PR or implemented the changes themselves, given that it was already approved? Asking the contributor to recreate their work seems unfair and reflects poorly on the process.

Great question! I don't know of a way to "migrate a PR." And for a languishing PR, it does not seem to be worth the effort. Some PRs are genuinely abandoned.

**Comment by devoncarew** [comment]
@Turskyi, thanks for the feedback.

We're transferring the issues shortly; see the last comment on https://github.com/dart-lang/linter/issues/4729. They won't be closed here and opened elsewhere, they'll just transfer to the sdk repo.

Re: the PR - the source of truth for the code for the linter moved over a year ago (on or around https://github.com/dart-lang/linter/commit/59cd18d1664915df1d1115cda84acda75675e1e2); after that point merging PRs here would no longer makes sense. Existing PRs should have been cleaned up then.

And I can't speak to this specific PR, but PRs can't be transferred like issues; a re-review of already approved content would likely be relatively brief.

**Comment by Turskyi** [comment]
Thank you for your transparency. However, the underlying message here seems to be: "_We’re okay with letting this PR go because fixing it ourselves isn’t worth the effort._" While I understand the challenges of repository migrations and triaging contributions, this approach risks alienating contributors who invest time and effort to improve the framework.
Open-source projects thrive on collaboration, and dismissing already-approved contributions sends the wrong message. I hope this can be a lesson to prioritize processes that respect community efforts and ensure their work doesn’t go to waste due to internal changes.

**Pull request closed without merging**
